core/client/src/main/java/alluxio/client/block/stream/LocalFilePacketWriter.java

```diff
@@ -0,0 +1,130 @@
+ */
+@NotThreadSafe
+public final class LocalFilePacketWriter implements PacketWriter {
+  private static final long PACKET_SIZE =
+      Configuration.getBytes(PropertyKey.USER_LOCAL_WRITER_PACKET_SIZE_BYTES);
+  private static final long FILE_BUFFER_BYTES =
+      Configuration.getBytes(PropertyKey.USER_FILE_BUFFER_BYTES);
+
+  /** The position to write the next byte at. */
+  private long mPos = 0;
+  /** The number of bytes reserved on the block worker to hold the block. */
+  private long mPosReserved = 0;
+  private final long mBlockId;
+  private final LocalFileBlockWriter mWriter;
+  private final BlockWorkerClient mBlockWorkerClient;
+  private boolean mClosed = false;
+
+  /**
+   * Creates an instance of {@link LocalFilePacketWriter}. This requires the block to be locked
+   * beforehand.
+   *
+   * @param blockWorkerClient the block worker client, not owned by this class
+   * @param blockId the block ID
+   * @throws IOException if it fails to create the packet writer
+   * @return the {@link LocalFilePacketWriter} created
+   */
+  public static LocalFilePacketWriter createLocalPacketWriter(BlockWorkerClient blockWorkerClient,
+      long blockId) throws IOException {
+    return new LocalFilePacketWriter(blockWorkerClient, blockId);
+  }
+
+  @Override
+  public long pos() {
+    return mPos;
+  }
+
+  @Override
+  public int packetSize() {
+    return (int) PACKET_SIZE;
+  }
+
+  @Override
+  public void writePacket(final ByteBuf buf) throws IOException {
+    try {
+      Preconditions.checkState(!mClosed, "PacketWriter is closed while writing packets.");
+      reserve(mPos + buf.readableBytes());
+      mPos += buf.readableBytes();
+      buf.readBytes(mWriter.getChannel(), buf.readableBytes());
+    } finally {
+      buf.release();
+    }
+  }
+
+  @Override
+  public void cancel() {}
+
+  @Override
+  public void flush() {}
+
+  @Override
+  public void close() throws IOException {
+    if (mClosed) {
+      return;
+    }
+    try {
+      mWriter.close();
+    } finally {
+      mClosed = true;
+    }
+  }
+
+  /**
+   * Creates an instance of {@link LocalFilePacketWriter}.
+   *
+   * @param blockWorkerClient the block worker client, not owned by this class
+   * @param blockId the block ID
+   * @throws IOException if it fails to create the packet writer
+   */
+  private LocalFilePacketWriter(BlockWorkerClient blockWorkerClient, long blockId)
+      throws IOException {
+    String blockPath = blockWorkerClient.requestBlockLocation(blockId, FILE_BUFFER_BYTES);
+    mWriter = new LocalFileBlockWriter(blockPath);
+    mPosReserved += FILE_BUFFER_BYTES;
+    mBlockId = blockId;
+    mBlockWorkerClient = blockWorkerClient;
+  }
+
+  /**
+   * Reserves enough space in the block worker.
+   *
+   * @param pos the pos of the file/block to reserve to
+   * @throws IOException if it fails to reserve the space
+   */
+  private void reserve(long pos) throws IOException {
```

Comment:
Can the argument be an amount? Using an absolute position doesn't seem to help because we always adjust the position when using `reserve`